### PR TITLE
BFD-2906: Automate Maven Release Process

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2906](https://jira.cms.gov/browse/BFD-2906)

**User Story or Bug Summary:**

As a BFD engineer I want to be able to have an automated release process that uses the maven release plugin to push build artifacts to CodeArtifact and ECR.

---

### What Does This PR Do?

**NOTE: This PR is being merged into an intermediate branch as additional changes are required before this process is usable**

This PR:

- Adds a new GitHub Actions workflow named `Build Release` that can be started manually (see [`workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch)). 
  - This workflow executes the following steps:
    - First, this new workflow will execute the Maven Release plugin, which will:
      - Build all modules within the parent `pom.xml`
      - Upon successful build, update _all_ `pom.xml`s to the version specified by the operator
      - Create a commit with these updated `pom.xml`s on the branch specified by the operator (`master`, in the typical case), and create a tag from this commit with the release version as the tag name
      - Update the `pom.xml`s _again_ setting their versions to the _development version_ specified by the operator (typically `<releaseVersion + 1>-SNAPSHOT`)
      - Create a commit with these _`SNAPSHOT`_ `pom.xml`s on the branch specified (usually `master`)
      - The Docker images generated by Jib will then be uploaded to ECR
      - All artifacts generated by the Maven Release build will then be uploaded to CodeArtifact under the `bfd-mgmt` repository, versioned appropriately
    - Then, assuming a successful run of the Maven Release plugin, the workflow will then generate a new GitHub prerelease based upon the Git tag generated by the Maven Release plugin with all of the relevant data dictionary files uploaded as assets under the new Release
  - Note that all Git operations done by this Workflow use the `bfd-release` GitHub Application, which is configured to _bypass_ branch protection rules on `master`
- Updates the `bfd-data-*` modules to be included as part of the parent `pom.xml` so that they are versioned and built similarly to all other modules

This PR has been validated by:

- Running the new `Build Release` workflow against a protected branch, verifying that the workflow runs successfully, all Git operations succeed (and protections are bypassed), all artifacts are uploaded to their respective repositories, and a corresponding GitHub Release is generated with the appropriate data dictionary files appended as assets
- `terraform plan`ning the changes to `mgmt`, verifying that the plan does not include any unexpected changes and that all changes are valid

No changes to our infrastructure have been made as of this PR. I will apply `mgmt` when this PR is merged.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [x] Yes
    - [ ] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- Branch protection rules on `master` are converted to equivalent rules using Rulesets, with the `bfd-release` GitHub Application exempted from those rulesets

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: .github/workflows/build-release.yml [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+          git config --global user.name "GitHub Actions"
+
+      - name: Generate maven toolchain config
+        run: |
+          cat << EOF > ~/.m2/toolchains.xml
+          <toolchains>
+            <toolchain>
+              <type>jdk</type>
+              <provides>
+                <version>17</version>
```
Comment (brandoncruz3) [inline .github/workflows/build-release.yml:88]: Just a note this will need to be updated to v21 soon ™️

**Review by brandoncruz3 — APPROVED** [review]
Looks good and verified in GHA. 🚀

**Review by StewGoin — APPROVED** [review]
After discussion with @mjburling today (I will allow him to summarize in a comment, etc...) I'm ok with this moving forward within the context provided by that forthcoming comment.

File: .github/workflows/build-release.yml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+              </server>
+            </servers>
+          </settings>
+          EOF
+
+      - name: "Prepare and Perform Release"
+        if: github.event_name == 'workflow_dispatch'
+        run: |-
+          mvn --batch-mode --activate-profiles test-release \
+              -Dtag="$BFD_RELEASE" \
```
Comment (brianburton) [inline .github/workflows/build-release.yml:140]: When do we plan to rename the maven profile?

**Comment by lsmitchell** [comment]
Is updating the documentation for our release process included in this story, or is it recorded somewhere?

File: .github/workflows/build-release.yml (hunk 2 above)
Comment (malessi) [reply .github/workflows/build-release.yml:140]: I'm good with renaming it in the follow-up Story, if that works with you? `test-release` obviously isn't ideal, but I don't think it being named `test-release` would block us from merging this now.

**Review by lsmitchell — APPROVED** [review]

**Review by brianburton — APPROVED** [review]
Excellent work!

**Comment by mjburling** [comment]
> After discussion with @mjburling today (I will allow him to summarize in a comment, etc...) I'm ok with this moving forward within the context provided by that forthcoming comment.

Very good! This change set, upon acceptance into the trunk (not necessarily _this_ feature branch that's targeting an intermediate branch), requires that the automation can be excluded from any rules that prohibit pushing to the protected `master` branch. While this could be achieved with a personal access token associated with an existing, trusted, real user as we do for our Jenkins-to-GitHub interactions, GitHub's branch protection rules **and* rulesets don't differentiate between token and key authentication, leaving the selected user open to potential inadvertent pushes to `master`. No thanks! Instead, we're electing to create a GitHub Application. At this time, it's my personally owned GitHub Application: [bfd-release](https://github.com/apps/bfd-release). This is _little_ different from the personal access token approach, but it helps us avoid:
- the risk of inadvertent pushes against `master`
- over-privileged tokens–`bfd-release` is only scoped to **just** this project

Ideally, this should be a GitHub Application at the CMSgov organization level. We're working through those details. In the meantime, we're moving ahead with this. **If this is still needed by the time we introduce these features into the trunk, this requires some light maintenance instructions for DASG personnel**

Also of note, what's **not** visible as part of this change set. This requires that the project adopts GitHub Rulesets to accommodate the `bfd-release` to help us avoid the over-application of branch protection rules for well-understood and safer automation practices. This also requires some additional documentation.

File: .github/workflows/build-release.yml (hunk 1 above)
Comment (mjburling) [reply .github/workflows/build-release.yml:88]: Necromancing this comment 💀. With #2066 having landed as of yesterday, I just want to make things a _little_ easier to reason about. Let's be sure that the target branch is up-to-date with the trunk and that **this** branch is up-to-date with the target.

To that end, can you open a draft PR for BFD-3078 and... make it so?

File: .github/workflows/build-release.yml [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        env:
+          CA_NAMESPACE: gov.cms.bfd
+          CA_PACKAGE: bfd-server-war
+          CA_DATA_DICTIONARY_ASSETS: |
+            [
+              "bfd-server-war-${{ inputs.releaseVersion }}-v1-data-dictionary.csv",
+              "bfd-server-war-${{ inputs.releaseVersion }}-v2-data-dictionary.csv",
+              "bfd-server-war-${{ inputs.releaseVersion }}-v1-data-dictionary.json",
+              "bfd-server-war-${{ inputs.releaseVersion }}-v2-data-dictionary.json"
+            ]
```
[the remark this answers is not in the document]
Comment (malessi) [reply .github/workflows/build-release.yml]: I've updated this to include the new `.xlsx` Excel worksheet and to follow the updated naming conventions introduced in #2065. That _should_ be enough, but let me know if I missed something.

**Pull request merged**